File: DominantColor/Shared/ColorSpaceConversion.m
//
//  ColorSpaceConversion.m
//  DominantColor
//
//  Created by Indragie on 12/21/14.
//  Copyright (c) 2014 Indragie Karunaratne. All rights reserved.
//

#import "ColorSpaceConversion.h"

#pragma mark - RGB

static GLKVector3 RGBToSRGB(GLKVector3 rgbVector) {
#if TARGET_OS_IPHONE
    // sRGB is the native device color space on iOS, no conversion is required.
    return rgbVector;
#else
    NSColor *rgbColor = [NSColor colorWithDeviceRed:rgbVector.x green:rgbVector.y blue:rgbVector.z alpha:1.0];
    NSColor *srgbColor = [rgbColor colorUsingColorSpace:NSColorSpace.sRGBColorSpace];
    return GLKVector3Make(srgbColor.redComponent, srgbColor.greenComponent, srgbColor.blueComponent);
#endif
}

static GLKVector3 SRGBToRGB(GLKVector3 srgbVector) {
#if TARGET_OS_IPHONE
    // sRGB is the native device color space on iOS, no conversion is required.
    return srgbVector;
#else
    const CGFloat components[4] = { srgbVector.x, srgbVector.y, srgbVector.z, 1.0 };
    NSColor *srgbColor = [NSColor colorWithColorSpace:NSColorSpace.sRGBColorSpace components:components count:4];
    NSColor *rgbColor = [srgbColor colorUsingColorSpace:NSColorSpace.deviceRGBColorSpace];
    return GLKVector3Make(rgbColor.redComponent, rgbColor.greenComponent, rgbColor.blueComponent);
#endif
}

#pragma mark - SRGB
// http://en.wikipedia.org/wiki/SRGB#Specification_of_the_transformation

static GLKVector3 SRGBToLinearSRGB(GLKVector3 srgbVector) {
    float (^f)(float) = ^float (float c) {
        if (c <= 0.04045f) {
            return c / 12.92f;
        } else {
            return powf((c + 0.055f) / 1.055f, 2.4f);
        }
    };
    return GLKVector3Make(f(srgbVector.x), f(srgbVector.y), f(srgbVector.z));
}

static GLKVector3 LinearSRGBToSRGB(GLKVector3 lSrgbVector) {
    float (^f)(float) = ^float (float c) {
        if (c <= 0.0031308f) {
            return c * 12.92f;
        } else {
            return (1.055f * powf(c, 1.f / 2.4f)) - 0.055f;
        }
    };
    return GLKVector3Make(f(lSrgbVector.x), f(lSrgbVector.y), f(lSrgbVector.z));
}

#pragma mark - XYZ (CIE 1931)
// http://en.wikipedia.org/wiki/CIE_1931_color_space#Construction_of_the_CIE_XYZ_color_space_from_the_Wright.E2.80.93Guild_data

static const GLKMatrix3 LinearSRGBToXYZMatrix = (GLKMatrix3){
    0.4124f, 0.2126f, 0.0193f,
    0.3576f, 0.7152f, 0.1192f,
    0.1805f, 0.0722f, 0.9505f
};

static GLKVector3 LinearSRGBToXYZ(GLKVector3 linearSrgbVector) {
    const GLKVector3 unscaledXYZVector = GLKMatrix3MultiplyVector3(LinearSRGBToXYZMatrix, linearSrgbVector);
    return GLKVector3MultiplyScalar(unscaledXYZVector, 100.f);
}

static const GLKMatrix3 XYZToLinearSRGBMatrix = (GLKMatrix3){
    3.2406f, -0.9689f, 0.0557f,
    -1.5372f, 1.8758f, -0.2040f,
    -0.4986f, 0.0415f, 1.0570f
};

static GLKVector3 XYZToLinearSRGB(GLKVector3 xyzVector) {
    const GLKVector3 scaledXYZVector = GLKVector3DivideScalar(xyzVector, 100.f);
    return GLKMatrix3MultiplyVector3(XYZToLinearSRGBMatrix, scaledXYZVector);
}

#pragma mark - LAB
// http://en.wikipedia.org/wiki/Lab_color_space#CIELAB-CIEXYZ_conversions

static GLKVector3 XYZToLAB(GLKVector3 xyzVector, GLKVector3 tristimulus) {
    float (^f)(float) = ^float (float t) {
        if (t > powf(6.f / 29.f, 3.f)) {
            return powf(t, 1.f / 3.f);
        } else {
            return ((1.f / 3.f) * powf(29.f / 6.f, 2.f) * t) + (4.f / 29.f);
        }
    };
    const float fx = f(xyzVector.x / tristimulus.x);
    const float fy = f(xyzVector.y / tristimulus.y);
    const float fz = f(xyzVector.z / tristimulus.z);
    
    const float l = (116.f * fy) - 16.f;
    const float a = 500 * (fx - fy);
    const float b = 200 * (fy - fz);
    
    return GLKVector3Make(l, a, b);
}

static GLKVector3 LABToXYZ(GLKVector3 labVector, GLKVector3 tristimulus) {
    float (^f)(float) = ^float (float t) {
        if (t > (6.f / 29.f)) {
            return powf(t, 3.f);
        } else {
            return 3.f * powf(6.f / 29.f, 2.f) * (t - (4.f / 29.f));
        }
    };
    const float c = (1.f / 116.f) * (labVector.x + 16.f);
    
    const float y = tristimulus.y * f(c);
    const float x = tristimulus.x * f(c + ((1.f / 500.f) * labVector.y));
    const float z = tristimulus.z * f(c - ((1.f / 200.f) * labVector.z));
    
    return GLKVector3Make(x, y, z);
}

#pragma mark - Public

// From http://www.easyrgb.com/index.php?X=MATH&H=15#text15
static const GLKVector3 D65Tristimulus = (GLKVector3){ 95.047f, 100.f, 108.883f };

INVector3 IN_RGBToLAB(INVector3 rgbVector) {
    const GLKVector3 gVector = INVector3ToGLKVector3(rgbVector);
    const GLKVector3 srgbVector = RGBToSRGB(gVector);
    const GLKVector3 lSrgbVector = SRGBToLinearSRGB(srgbVector);
    const GLKVector3 xyzVector = LinearSRGBToXYZ(lSrgbVector);
    const GLKVector3 labVector = XYZToLAB(xyzVector, D65Tristimulus);
    return GLKVector3ToINVector3(labVector);
}

INVector3 IN_LABToRGB(INVector3 labVector) {
    const GLKVector3 gVector = INVector3ToGLKVector3(labVector);
    const GLKVector3 xyzVector = LABToXYZ(gVector, D65Tristimulus);
    const GLKVector3 lSrgbVector = XYZToLinearSRGB(xyzVector);
    const GLKVector3 srgbVector = LinearSRGBToSRGB(lSrgbVector);
    const GLKVector3 rgbVector = SRGBToRGB(srgbVector);
    return GLKVector3ToINVector3(rgbVector);
}

File: DominantColor/Shared/ColorSpaceConversion.swift
//
//  ColorSpaceConversion.swift
//  DominantColor
//
//  Created by Jernej Strasner on 2/5/19.
//  Copyright © 2019 Indragie Karunaratne. All rights reserved.
//

#if os(iOS)
import UIKit
#elseif os(OSX)
import AppKit
#endif

import simd

// MARK: - RGB

func RGBToSRGB(_ rgbVector: simd_float3) -> simd_float3 {
    #if os(iOS)
    return rgbVector
    #elseif os(OSX)
    let rgbColor = NSColor(deviceRed: CGFloat(rgbVector.x), green: CGFloat(rgbVector.y), blue: CGFloat(rgbVector.z), alpha: 1.0)
    guard let srgbColor = rgbColor.usingColorSpace(.sRGB) else {
        fatalError("Could not convert color space")
    }
    return simd_float3(Float(srgbColor.redComponent), Float(srgbColor.greenComponent), Float(srgbColor.blueComponent))
    #endif
}

func SRGBToRGB(_ srgbVector: simd_float3) -> simd_float3 {
    #if os(iOS)
    return srgbVector
    #elseif os(OSX)
    let components: [CGFloat] = [CGFloat(srgbVector.x), CGFloat(srgbVector.y), CGFloat(srgbVector.z), 1.0]
    let srgbColor = NSColor(colorSpace: .sRGB, components: components, count: 4)
    guard let rgbColor = srgbColor.usingColorSpace(.deviceRGB) else {
        fatalError("Could not convert color space")
    }
    return simd_float3(Float(rgbColor.redComponent), Float(rgbColor.greenComponent), Float(rgbColor.blueComponent))
    #endif
}

// MARK: - SRGB

func SRGBToLinearSRGB(_ srgbVector: simd_float3) -> simd_float3 {
    func f(_ c: Float) -> Float {
        if (c <= 0.04045) {
            return c / 12.92
        } else {
            return powf((c + 0.055) / 1.055, 2.4)
        }
    }
    return simd_float3(f(srgbVector.x), f(srgbVector.y), f(srgbVector.z))
}

func LinearSRGBToSRGB(_ lSrgbVector: simd_float3) -> simd_float3 {
    func f(_ c: Float) -> Float {
        if (c <= 0.0031308) {
            return c * 12.92
        } else {
            return (1.055 * powf(c, 1.0 / 2.4)) - 0.055
        }
    };
    return simd_float3(f(lSrgbVector.x), f(lSrgbVector.y), f(lSrgbVector.z));
}

// MARK: - XYZ (CIE 1931)
// http://en.wikipedia.org/wiki/CIE_1931_color_space#Construction_of_the_CIE_XYZ_color_space_from_the_Wright.E2.80.93Guild_data

let LinearSRGBToXYZMatrix = simd_float3x3([
    SIMD3(0.4124, 0.2126, 0.0193),
    SIMD3(0.3576, 0.7152, 0.1192),
    SIMD3(0.1805, 0.0722, 0.9505)
])

func LinearSRGBToXYZ(_ linearSrgbVector: simd_float3) -> simd_float3 {
    let unscaledXYZVector = LinearSRGBToXYZMatrix * linearSrgbVector
    return unscaledXYZVector * 100.0
}

let XYZToLinearSRGBMatrix = simd_float3x3([
    SIMD3(3.2406, -0.9689, 0.0557),
    SIMD3(-1.5372, 1.8758, -0.2040),
    SIMD3(-0.4986, 0.0415, 1.0570)
])

func XYZToLinearSRGB(_ xyzVector: simd_float3) -> simd_float3 {
    let scaledXYZVector = xyzVector / 100.0
    return XYZToLinearSRGBMatrix * scaledXYZVector
}


// MARK: - LAB
// http://en.wikipedia.org/wiki/Lab_color_space#CIELAB-CIEXYZ_conversions

func XYZToLAB(_ xyzVector: simd_float3, _ tristimulus: simd_float3) -> simd_float3 {
    func f(_ t: Float) -> Float {
        if (t > powf(6.0 / 29.0, 3.0)) {
            return powf(t, 1.0 / 3.0)
        } else {
            return ((1.0 / 3.0) * powf(29.0 / 6.0, 2.0) * t) + (4.0 / 29.0)
        }
    };
    let fx = f(xyzVector.x / tristimulus.x)
    let fy = f(xyzVector.y / tristimulus.y)
    let fz = f(xyzVector.z / tristimulus.z)

    let l = (116.0 * fy) - 16.0
    let a = 500 * (fx - fy)
    let b = 200 * (fy - fz)

    return simd_float3(l, a, b)
}

func LABToXYZ(_ labVector: simd_float3, _ tristimulus: simd_float3) -> simd_float3 {
    func f(_ t: Float) -> Float {
        if (t > (6.0 / 29.0)) {
            return powf(t, 3.0)
        } else {
            return 3.0 * powf(6.0 / 29.0, 2.0) * (t - (4.0 / 29.0))
        }
    };
    let c = (1.0 / 116.0) * (labVector.x + 16.0)

    let y = tristimulus.y * f(c)
    let x = tristimulus.x * f(c + ((1.0 / 500.0) * labVector.y))
    let z = tristimulus.z * f(c - ((1.0 / 200.0) * labVector.z))

    return simd_float3(x, y, z)
}

// MARK: - Public

// From http://www.easyrgb.com/index.php?X=MATH&H=15#text15
let D65Tristimulus = simd_float3(5.047, 100.0, 108.883)

func IN_RGBToLAB(_ gVector: simd_float3) -> simd_float3 {
    let srgbVector = RGBToSRGB(gVector)
    let lSrgbVector = SRGBToLinearSRGB(srgbVector)
    let xyzVector = LinearSRGBToXYZ(lSrgbVector)
    let labVector = XYZToLAB(xyzVector, D65Tristimulus)
    return labVector
}

func IN_LABToRGB(_ gVector: simd_float3) -> simd_float3 {
    let xyzVector = LABToXYZ(gVector, D65Tristimulus)
    let lSrgbVector = XYZToLinearSRGB(xyzVector)
    let srgbVector = LinearSRGBToSRGB(lSrgbVector)
    let rgbVector = SRGBToRGB(srgbVector)
    return rgbVector
}
